How to write good tests for parallel environments (#2182)
scality/Zenko: tests/ctst/HOW_TO_WRITE_TESTS.md

Reviewer: should this instead go in a more "natural" location, e.g. `docs/writing-ctst-tests.md` ?

# How to write a test that runs in parallel

> Writting intergation / behavior tests that run in parallel is not trivial.

Reviewer 2: ```suggestion
Writing integration / behavior tests that runs in parallel is not trivial.
```

This page lists all the rules to follow when writing a test that runs in
parallel with other tests.

*This is specific to Zenko, but the rules are applicable for any suite of*
*tests in parallel.*

## 1. Tests must be idempotent.

In parallel, the tests can run in any order. It's not because a test works at a
given time, that it will work if the test order changes after adding new ones
or renaming files. You must ensure the scenarios are fully idempotent.

> "Flakies" are mostly due to non-idempotent tests, then bugs. It is rarely due

Reviewer 2: > it is rarely due to the platform

Even though it's *actually* true, I'm not sure we want to write that in the documentation, as the test are made to verify that the platform works..

Author: By platform I mean the CI environment, I will make it clearer that we cannot always think it's due to limited resources in the CI, or some sort of slowness we cannot explain

# How to write a test that runs in parallel

Writting intergation / behavior tests that run in parallel is not trivial.
This page lists all the rules to follow when writing a test that runs in
parallel with other tests.

*This is specific to Zenko, but the rules are applicable for any suite of*
*tests in parallel.*

**Rule #1**

> > Tests must be idempotent.

Reviewer: ```suggestion
> Tests should be idempotent.
```

Author: Why "should" and not "must"? it is not supposed to be a suggestion.

In parallel, the tests can run in any order. It's not because a test works at a
given time, that it will work if the test order changes after adding new ones
or renaming files. You must ensure the scenarios are fully idempotent.
"Flakies" are mostly due to non-idempotent tests, then bugs. It is rarely due
to the platform.

**Rule #2**

> Use unique resources names. S3 Bucket, Account and S3 Objects should be
> unique.

It limits the risks of collision between tests.
For cross-account scenarios, use unique S3 Policies and S3 Roles.

If data is used, such as lock files, ensure they are named appropriately.

**Rule #3**

> Avoid reconfiguring the environment during test execution.

If a reconfiguration triggers a reconciliation of the resources, running tests
might be affected. Also, it restarts the services, which leads to missing logs,
harder debugging, and longer test execution.

Examples:

- Location creation;
- Bucket Website endpoints.

> ... must be created before the test start, or in a dedicated set of test.

Reviewer: ```suggestion
... should be created before the test start, or in a dedicated set of test.
```

> ... must be created before the test start, or in a dedicated set of test.

Reviewer: > or in a dedicated set of test

not clear what you mean by "set of test" here : a separate test suite, a separate execution of ctst , ... ?
having separate executions should be avoided, as it makes it more likely that not all tested are executed (when developping locally, once we can do it easily...)

**Rule #4**

> Do not use `atMostOnePicklePerTag`.

If a set of scenario requires the use of `atMostOnePicklePerTag`, they won't
be executed in parallel, which is:

- Not realistic for the production environment.
- Not efficient for the test execution: the duration of tests will suffer.
- A proof that the test is not following **Rule #1**.

Ensuring this rule may be hard, and idempotent scenarios are not always
possible. Solutions exist:

- Pre-create the resources required for the tests before they start.
- Use `Before` / `After` hooks to setup your environment once.
  Concurrency can be controlled with lock files.
- Make your assertions more flexible: strict and absolute checks are good
  for unit or functional testing, but in integration tests, prefer using
  relative checks.
- As a last resort, we might have a dedicated test suite.

**Rule #5**

> > Avoid scenarios with no value.

Reviewer: i think the last line of the description is a better name for the rule ("no value" is not an appropriate description: doing many tests in zenko still has more value than no tests at all...)

```suggestion
> Focus on validating features
```

It is tempting to write scenarios that test more than what is needed

> at the integration testing layer: please trust your unit and functional

Reviewer: > integration testing layer

this is not really "integration tests" : as far as semantics go, these tests are more akin to "acceptance tests" and "smoke tests".... which we rely on to validate that components are properly integrated.
Best rephrase to avoid the confusion IMHO.

> please trust your unit and functional

"trust" is probably not what we want : if we really trusted, we wouldn't do any test :-)
```
please use unit and functional tests to fully validate the behavior of each component;
and e2e tests only to ensure that the product-level feature is working as expected per
the requirements. Making good use of unit and functional tests makes testing all cases,
including corner cases or error handling, must easier, leading to higher quality; and
improves velocity by having a much faster feedback loop.
```

tests. If a scenario has too many variants, then likely they can be tested
in lower layers. Please focus on validating a feature.

**Rule #6**

> Ensure assertions are idempotent after sleep times.

Some tests might need to "sleep" for some time, either static or using
more complex logic. The subsequent assertions must take into account
any operation on the platform that might be asynchronous.

Example: we run a Kubernetes job in tests that look idempotent. However,
the job itself might not be idempotent and run in parallel, leading to
some resources being skipped by a previous job that would take longer than

> expected, and the test results will differ from what is expected.

Reviewer: not sure the exemple looks so clear, for an "outside" eye with no content on which job and what they do...

**Rule #7**

> Avoid using the @Flaky tag.

> If a test is marked as flaky, it is likely not idempotent, or we did not
> do our job correctly to understand the problem. Please fix the test.
> 
> If an issue exists in the product, prefer fixing the bug first.

Reviewer: ```suggestion
Whenever there is a flaky, there is an issue : either a bug (corner case or race condition) in the product,
or simply an issue in the test. So the issue should be fixed either way.

More often than not, retrying the build (i.e. ignoring the flaky) or automating the retry with @Flaky is just
masking a bug, and leaving it in production instead of fixing it : so it must not be done.

One frequent problem in flaky tests is that they are not idempotent, or that we did not do our job correctly to understand the problem. Flaky tests must be investigated, and issues in the tests must be fixed to ensure we
don't ignore an actual problem in the product.
```

**Rule #8**

> Teardown resources.

If a test creates resources, ensure they are deleted at the end of the test.
Do not delete the resources in case of error, to ease debugging.

**Rule #9**

> Avoid tests relying on other tests execution.

It differs from **Rule #1** in the sense that a test might be idempotent but
rely on the execution order or the result of another set of tests.

Tests requiring to run at the very end of a parallel test suite with a specific
naming (e.g., `zzz.*`) should be removed. Prefer using `After` hooks with
specific tags.

**Rule #10**

> Validate your test before merging.

> If a test pass once, please ensure it passes for the good reasons.

Reviewer: ```suggestion
If a test passes once, please ensure it passes for the good reasons.
```

# How to write a test that runs in parallel

Writting intergation / behavior tests that run in parallel is not trivial.
This page lists all the rules to follow when writing a test that runs in
parallel with other tests.

*This is specific to Zenko, but the rules are applicable for any suite of*
*tests in parallel.*

> **Rule #1**

Reviewer 3: Just a general formatting remark: it would be more straightforward for readers to just name the rules as titles prefixed by their number. I find the bold rule number followed by the rule itself as a quote a bit distracting.

> Tests must be idempotent.

In parallel, the tests can run in any order. It's not because a test works at a
given time, that it will work if the test order changes after adding new ones
or renaming files. You must ensure the scenarios are fully idempotent.
"Flakies" are mostly due to non-idempotent tests, then bugs. It is rarely due
to the platform.

**Rule #2**

> Use unique resources names. S3 Bucket, Account and S3 Objects should be
> unique.

It limits the risks of collision between tests.
For cross-account scenarios, use unique S3 Policies and S3 Roles.

If data is used, such as lock files, ensure they are named appropriately.

**Rule #3**

> Avoid reconfiguring the environment during test execution.

If a reconfiguration triggers a reconciliation of the resources, running tests
might be affected. Also, it restarts the services, which leads to missing logs,
harder debugging, and longer test execution.

Examples:

- Location creation;
- Bucket Website endpoints.

... must be created before the test start, or in a dedicated set of tests,
that is, a set of test executed before all tests having a specific tag

> used for identifying the feature(s).

Reviewer 3: Unless we're explicitly testing the reconfiguration itself which we should probably do.

Also, running tests should not be affected, if they are it means that production will, when an upgrade or config change is made.

I agree with the debuggability/log availability part, but they're issues that need to be addressed, not really a reason to avoid reconfiguring altogether (provided of course that the reconfiguration does not touch other tests' resources).

**Rule #4**

> Do not use `atMostOnePicklePerTag`.

If a set of scenario requires the use of `atMostOnePicklePerTag`, they won't
be executed in parallel, which is:

- Not realistic for the production environment.
- Not efficient for the test execution: the duration of tests will suffer.
- A proof that the test is not following **Rule #1**.

Ensuring this rule may be hard, and idempotent scenarios are not always
possible. Solutions exist:

- Pre-create the resources required for the tests before they start.
- Use `Before` / `After` hooks to setup your environment once.
  Concurrency can be controlled with lock files.
- Make your assertions more flexible: strict and absolute checks are good
  for unit or functional testing, but in integration tests, prefer using
  relative checks.
- As a last resort, we might have a dedicated test suite.

**Rule #5**

> Focus on validating features.

It is tempting to write scenarios that test more than what is needed
at the acceptance testing layer: please use unit and functional tests to fully
validate the behavior of each component; and end-to-end tests only to ensure

> that the product-level feature is working as expected per the requirements.

Reviewer 3: Maybe it would be beneficial to clearly spell out that we only want to assert against externally visible state (as given in the feature's requirements) and NOT rely on internal state introspection, which end-users don't really care about and subject to change.

improves velocity by having a much faster feedback loop.

**Rule #6**

> Ensure assertions are idempotent after sleep times.

Some tests might need to "sleep" for some time, either static or using
more complex logic. The subsequent assertions must take into account
any operation on the platform that might be asynchronous and take place
during the sleep time.

**Rule #7**

> Avoid using the @Flaky tag.

Whenever there is a flaky, there is an issue : either a bug (corner case or
race condition) in the product, or simply an issue in the test. So the
issue should be fixed either way. More often than not, retrying the build
(i.e. ignoring the flaky) or automating the retry with @Flaky is just
masking a bug, and leaving it in production instead of fixing it : so it
must not be done. One frequent problem in flaky tests is that they are not
idempotent, or that we did not do our job correctly to understand the
problem. Flaky tests must be investigated, and issues in the tests must be
fixed to ensure we don't ignore an actual problem in the product.

**Rule #8**

> Teardown resources.

If a test creates resources, ensure they are deleted at the end of the test.

> Do not delete the resources in case of error, to ease debugging.

Reviewer 3: I both agree and disagree: tearing down resources should always be done and be reliable, and also we need to be able to debug. A middle ground would be to skip teardown on failure only if in debug mode (the one with tmate active), otherwise always do it.

This would make rule #9 redundant as tests relying on other tests to have run would just not work.

Author: Thank you for all the feedback I updated the page accordingly - I agree with everything but this on. I updated as you suggested though.

I think we should not cleanup resources in case of error even if we're not running debug. The reason is that as each test uses a partitioned set of resources (bucket/objects, possibly account), not cleaning in case of error (or flaky) helps us finding the root cause much more easily, by having access to the metadata ind the mongodb dumps, as well as the logs of all operations done by CTST for the failing test (identifiable with a single id). The overhead is minimal and it should not affect other tests in any way, even if the bucket is still here.
